File: dungeons-1.21.11/src/main/java/fishmod/utils/events/EventHandler.java
package fishmod.utils.events;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class EventHandler<T> {
    /**
     * This class handles some specified Event
     * to cancel the event if implemented return true
     * else return false
     */

    private final List<T> listeners = new ArrayList<>();

    public void register(T listener) {
        listeners.add(listener);
    }

    public boolean invoke(Predicate<T> action) {
        if (listeners.isEmpty()) return false;
        for (T listener : listeners) {
            if (action.test(listener)) {
                return true;
            }
        }
        return false;
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/events/EventHandler.kt
package fishmod.utils.events

import java.util.function.Predicate

/** Registered listeners for an event; a listener returning true cancels the event. */
class EventHandler<T> {

    private val listeners = ArrayList<T>()

    fun register(listener: T) {
        listeners.add(listener)
    }

    fun invoke(action: Predicate<T>): Boolean {
        if (listeners.isEmpty()) return false
        for (listener in listeners) {
            if (action.test(listener)) {
                return true
            }
        }
        return false
    }
}
